xdsl/backend/riscv/prologue_epilogue_insertion.py

```diff
@@ -0,0 +1,92 @@
+from xdsl.builder import Builder
+from xdsl.dialects import builtin, riscv, riscv_func
+from xdsl.dialects.riscv import (
+    FloatRegisterType,
+    IntRegisterType,
+    Registers,
+    RISCVRegisterType,
+)
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+
+
+class PrologueEpilogueInsertion(ModulePass):
+    """
+    Pass inserting a prologue and epilogue according to the RISC-V ABI.
+    The prologues and epilogues are responsible for saving any callee-preserved
+    registers.
+    In RISC-V these are 's0' to 's11' and 'fs0' to `fs11'.
+    The stack pointer 'sp' must also be restored to its original value.
+
+    This pass should be run late in the pipeline after register allocation.
+    It does not itself require register allocation nor invalidate the result of the
+    register allocator.
+    """
+
+    name = "riscv-prologue-epilogue-insertion"
+
+    def _process_function(self, func: riscv_func.FuncOp) -> None:
+        # Find all callee-preserved registers that are clobbered. We define clobbered
+        # as it being the result of some operation and therefore written to.
+        used_callee_preserved_registers = dict[
+            IntRegisterType | FloatRegisterType, None
+        ]()
+        for op in func.walk():
+            # Special cases ops that do not actually write to the register.
+            if isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp):
+                continue
+
+            for res in op.results:
+                if res.type in Registers.S or res.type in Registers.FS:
+                    used_callee_preserved_registers[res.type] = None
+
+        # Note: This assumes RV32D aka 4 byte integer registers, 8 byte FP registers.
+        def get_register_size(r: RISCVRegisterType):
+            if isinstance(r, IntRegisterType):
+                return 4
+            return 8
+
+        # Build the prologue at the beginning of the function.
+        builder = Builder.at_start(func.body.blocks[0])
+        sp_register = builder.insert(riscv.GetRegisterOp(Registers.SP))
+        stack_size = sum(get_register_size(r) for r in used_callee_preserved_registers)
+        builder.insert(riscv.AddiOp(sp_register, -stack_size, rd=Registers.SP))
+        offset = 0
+        for reg in used_callee_preserved_registers.keys():
+            if isinstance(reg, IntRegisterType):
+                reg_op = builder.insert(riscv.GetRegisterOp(reg))
+                op = riscv.SwOp(rs1=sp_register, rs2=reg_op, immediate=offset)
+            else:
+                reg_op = builder.insert(riscv.GetFloatRegisterOp(reg))
+                op = riscv.FSdOp(rs1=sp_register, rs2=reg_op, immediate=offset)
+
+            builder.insert(op)
+            offset += get_register_size(reg)
+
+        # Now build the epilogue right before every return operation.
```

Comment:
Is there not a requirement of a single exit block?

Reply:
There is not, see the tests + any of the kernels they generate. They contain multiple blocks and each block may have their own return operation. (You often even want this for hot-cold basic block splitting).
Given how trivial it is support this I don't see a reason not to support it.

Reply:
Yep sounds good, thanks for the clarification